### PR TITLE
[filesystem_device]Should add some checkpoints after detaching

### PR DESCRIPTION
**Pull request by iccaszhulili**
1) Should check the guest after detaching
2) Should check whether there is virtiofs after detaching

Signed-off-by: Lili Zhu <lizhu@redhat.com>

### CODE REVIEW

**Comment by iccaszhulili** [comment]
avocado run --vt-type libvirt filesystem_device
.... 
(266/269) type_specific.io-github-autotest-libvirt.virtual_devices.filesystem_device.negative_test.invalid_queue_size.not_powerof_2.hotplug_unplug.detach_device.fs_test.xattr_on.flock_on.lock_posix_on.cache_mode_none.one_fs.one_guest: PASS (22.60 s)
 (267/269) type_specific.io-github-autotest-libvirt.virtual_devices.filesystem_device.negative_test.invalid_queue_size.larger_than_1024.nop.fs_test.xattr_on.flock_on.lock_posix_on.cache_mode_none.one_fs.one_guest: PASS (8.05 s)
 (268/269) type_specific.io-github-autotest-libvirt.virtual_devices.filesystem_device.negative_test.invalid_queue_size.larger_than_1024.hotplug_unplug.detach_device.fs_test.xattr_on.flock_on.lock_posix_on.cache_mode_none.one_fs.one_guest: PASS (22.60 s)
 (269/269) type_specific.io-github-autotest-libvirt.virtual_devices.filesystem_device.negative_test.managedsave.nop.fs_test.xattr_on.flock_on.lock_posix_on.cache_mode_none.one_fs.one_guest: PASS (22.88 s)
RESULTS    : PASS 269 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 10166.84 s

**Review by chunfuwen — APPROVED** [review]
LGTM

File: libvirt/tests/src/virtual_device/filesystem_device.py [hunk 1]
```diff
@@ -350,14 +367,17 @@ def umount_fs(vm):
...
+                        if detach_device_alias:
+                            alias = fs_dev.alias['name']
+                            ret = virsh.detach_device_alias(vm.name, alias,
+                                                            ignore_status=True, debug=True)
+                        else:
+                            ret = virsh.detach_device(vm.name, fs_dev.xml,
+                                                      ignore_status=True, debug=True)
+                        libvirt.check_exit_status(ret, status_error)
+                        check_filesystem_in_guest(vm, fs_dev)
+                    check_detached_xml(vm)
```
Comment (Yingshun) [inline libvirt/tests/src/virtual_device/filesystem_device.py:380]: I have some questions about these lines:
1. The return values will be checked by libvirt.check_exit_status(ret, status_error) and  status_error is configurable. I wonder if check_filesystem_in_guest and check_detached_xml should be checked when status_error == False.
2. If status_error is always True, should we add wait_for_event=True to  virsh.detach_device and virsh.detach_device_alias?

File: libvirt/tests/src/virtual_device/filesystem_device.py [hunk 2]
```diff
@@ -151,6 +151,23 @@ def umount_fs(vm):
...
             session.close()
 
+    def check_detached_xml(vm):
+        vmxml = vm_xml.VMXML.new_from_dumpxml(vm.name)
+        filesystems = vmxml.devices.by_device_tag('filesystem')
+        if filesystems:
+            test.fail("There should be no filesystem devices in guest "
+                      "xml after hotunplug")
+
+    def check_filesystem_in_guest(vm, fs_dev):
```
Comment (Yingshun) [inline libvirt/tests/src/virtual_device/filesystem_device.py:161]: please add docstring.

File: libvirt/tests/src/virtual_device/filesystem_device.py [hunk 3]
```diff
@@ -151,6 +151,23 @@ def umount_fs(vm):
                 session.cmd('rm -rf %s' % mount_dir, ignore_all_errors=True)
             session.close()
 
+    def check_detached_xml(vm):
```
Comment (Yingshun) [inline libvirt/tests/src/virtual_device/filesystem_device.py:154]: docstring plz.

File: libvirt/tests/src/virtual_device/filesystem_device.py (hunk 2 above)
Comment (iccaszhulili) [reply libvirt/tests/src/virtual_device/filesystem_device.py:167]: Added. Thanks

File: libvirt/tests/src/virtual_device/filesystem_device.py (hunk 3 above)
Comment (iccaszhulili) [reply libvirt/tests/src/virtual_device/filesystem_device.py:154]: Added. Thanks

File: libvirt/tests/src/virtual_device/filesystem_device.py (hunk 1 above)
Comment (iccaszhulili) [reply libvirt/tests/src/virtual_device/filesystem_device.py]: @Yingshun 
1) Just to confirm, I thought your question might be " I wonder if check_filesystem_in_guest and check_detached_xml should be checked when status_error == **True**." If it is, I add "return" after attach_device execution. So there is no detach_device  exection if status_error=True.        
 if hotplug_unplug:
                for fs_dev in fs_devs:
                    ret = virsh.attach_device(vm_names[index], fs_dev.xml,
                                              ignore_status=True, debug=True)
                    libvirt.check_exit_status(ret, status_error)
                if status_error:
                    return
2) Updated. Thanks

**Review by Yingshun — APPROVED** [review]

**Pull request merged**